security/advancedtls/advancedtls.go

```diff
@@ -79,21 +81,64 @@ type GetRootCAsResults struct {
 	TrustCerts *x509.CertPool
 }
 
-// RootCertificateOptions contains a field and a function for obtaining root
-// trust certificates.
-// It is used by both ClientOptions and ServerOptions.
-// If users want to use default verification, but did not provide a valid
-// RootCertificateOptions, we use the system default trust certificates.
+// RootCertificateOptions contains options to obtain root trust certificates
+// for both the client and the server.
+// At most one of the options could be set. If none of them are set, we
+// use the system default trust certificates.
 type RootCertificateOptions struct {
-	// If field RootCACerts is set, field GetRootCAs will be ignored. RootCACerts
-	// will be used every time when verifying the peer certificates, without
-	// performing root certificate reloading.
+	// If RootCACerts is set, it will be used every time when verifying
+	// the peer certificates, without performing root certificate reloading.
 	RootCACerts *x509.CertPool
-	// If GetRootCAs is set and RootCACerts is nil, GetRootCAs will be invoked
-	// every time asked to check certificates sent from the server when a new
-	// connection is established.
-	// This is known as root CA certificate reloading.
-	GetRootCAs func(params *GetRootCAsParams) (*GetRootCAsResults, error)
+	// If GetRootCertificates is set, it will be invoked to obtain root certs in
+	// every new connection.
+	GetRootCertificates func(params *GetRootCAsParams) (*GetRootCAsResults, error)
+	// If RootProvider is set, we will use the root certs from the Provider's
+	// KeyMaterial() call in every new connection.
+	RootProvider certprovider.Provider
+}
+
+// nonNilFieldCount returns the number of set fields in RootCertificateOptions.
```

Comment:
Is this really required?
Why is clearly documenting which field takes priority when more than one field is set, not enough?

Reply:
Because in my opinion, there is no use case when users want to specify more than one field.  If users specify more than one field, it is most likely that they misunderstand some part of the API, and we probably want to point it out directly instead of proceeding with the implicit priorities.

Reply:
SG. Let's use the reflection to count non-nil fields and do the enforcement check for now. We can revisit this in the future if we have some real use cases that users have to specify more than one field.